application/controllers/ai.py

```diff
@@ -1,25 +1,228 @@
+    if card.giveMilitary > 0:
+        for n in neighbours:
+            if player.military - n.military > card.giveMilitary:
+                #your military is too high to be worth adding more
+                score += 0
+            elif player.military - n.military > 0:
+                #your military is high, but still worth reinforcing
+                score += card.giveMilitary*military_weight
+            elif player.military - n.military > -card.giveMilitary:
+                #getting this card will win you the battle!
+                score += 2*card.giveMilitary*military_weight
+            elif player.military - n.military > -2*card.giveMilitary:
+                #this card won't be enough, but might be worth trying to get there?
+                score += card.giveMilitary*military_weight
 
-    # Add bonuses if resources needed for wonder -- TODO
-    wonder_cards = get_all_wonder_cards(player)
+    # wonder_cards = get_all_wonder_cards(player)
 
     # Add bonus for lots of research cards
+    if card.colour == 'green':
+        #bonus to get ai to build sicence at all
+        if (player.glass + player.paper + player.cloth >= 2 or player.advanced_cheap_trade) and card.age == 1:
+            score += 4*research_weight
+        #number of research cards increases viability for more
+        score += (player.cog + player.tablet + player.compass + player.wildcard)*research_weight
+        #preference for complete set > biggest symbol > other
+        if card.giveResearch == 'cog':
+            if player.cog < player.tablet and player.cog < player.compass:
+                score *= 1 + 2*set_weight
+            elif player.cog > player.tablet and player.cog > player.compass:
+                score *= 1 + set_weight
+        if card.giveResearch == 'tablet':
+            if player.tablet < player.cog and player.tablet < player.compass:
+                score *= 1 + 2*set_weight
+            elif player.tablet > player.cog and player.tablet > player.compass:
+                score *= 1 + set_weight
+        if card.giveResearch == 'compass':
+            if player.compass < player.tablet and player.compass < player.cog:
+                score *= 1 + 2*set_weight
+            elif player.compass > player.tablet and player.compass > player.cog:
+                score *= 1 + set_weight
+    elif card.name == 'Scientists Guild':
+        #more appropriate here than in guild section
+        if player.glass + player.paper + player.cloth + int(player.advanced_cheap_trade) >= 2:
+            score += 3*research_weight
+        score += (player.cog + player.tablet + player.compass + player.wildcard)*research_weight
+        score += 1 + 2*set_weight #always fits in best spot
 
     # Yellow/purple card treatment
+    
+    score += card.giveMoney
+    if card.giveMoney and player.money < 2*card.age:
+        #bonus when low on money
+        score += 2*card.age - player.money
+    
+    
+    if card.name == 'Marketplace':

```

Comment:
I would split this if statement in two, one for yellow and purple cards, then before entering the if statement, check if the card is that colour. Currently, every card goes through 18 predicate checks when they may not even be a yellow or purple card.